contracts/smart-account/utils/AddressResolver.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {IEcdsaOwnershipRegistryModule} from "../interfaces/modules/IEcdsaOwnershipRegistryModule.sol";
+import {IAddressResolver} from "../interfaces/IAddressResolver.sol";
+import {ISmartAccount} from "../interfaces/ISmartAccount.sol";
+
+/// EOA <-> Smart Account address resolver for Biconomy smart accounts
+contract AddressResolver {
+    struct SmartAccountResult {
+        address accountAddress;
+        address factoryAddress;
+        address currentImplementation;
+        string currentVersion;
+        string factoryVersion;
+        uint256 deploymentIndex;
+    }
+
+    address public constant SA_V1_FACTORY =
+        0x000000F9eE1842Bb72F6BBDD75E6D3d4e3e9594C;
+    address public constant SA_V2_FACTORY =
+        0x000000a56Aaca3e9a4C479ea6b6CD0DbcB6634F5;
+    address public constant ECDSA_REGISTRY_MODULE_ADDRESS =
+        0x0000001c5b32F37F5beA87BDD5374eB2aC54eA8e;
+
+    /**
+     * @dev Returns the addresses of all the smart accounts deployed by the EOA for any deployment index from 0 to _maxIndex.
+     * @param _eoa Address of the EOA.
+     * @param _maxIndex Maximum index to check.
+     * @notice This function is only for V1 and V2 Biconomy smart accounts.
+     * @notice For V2 smart accounts, the _moduleAddress and _moduleSetupData parameters are not used. It assumes ECDSA module.
+     */
+    function resolveAddresses(
+        address _eoa,
+        uint8 _maxIndex
+    ) external view returns (SmartAccountResult[] memory) {
+        SmartAccountResult[] memory _saInfo = new SmartAccountResult[](
+            _maxIndex * 2
+        );
+        uint256 saInfoIndex = 0; // To keep track of the current index in _saAddresses
+
+        for (uint256 i; i < _maxIndex; i++) {
+            address v1Address = IAddressResolver(SA_V1_FACTORY)
+                .getAddressForCounterFactualAccount(_eoa, i);
+            if (v1Address != address(0) && _isSmartContract(v1Address)) {
```

Comment:
address(0) will be checked first.  
I thought CREATE2 would return 0 address but I see that's not the case since this is computing without using opcode. 
so read call and anyway for read call this should not matter